[Cocoa] Implement LinearMediaKit versions of PlaybackSessionInterface and VideoPresentationInterface

aestes opened the pull request:
#### fecabc2385988b2c0e4ba5397bf8d6c3f9e5721a
<pre>
[Cocoa] Implement LinearMediaKit versions of PlaybackSessionInterface and VideoPresentationInterface
<a href="https://bugs.webkit.org/show_bug.cgi?id=269442">https://bugs.webkit.org/show_bug.cgi?id=269442</a>
<a href="https://rdar.apple.com/problem/122995773">rdar://problem/122995773</a>

Reviewed by Jer Noble.

Provided implementations of PlaybackSessionInterfaceLMK and VideoPresentationInterfaceLMK backed by
LMPlayableViewController and WKSLinearMediaPlayer.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
Implemented toggleFullscreen by setting the media element&apos;s presentation mode.

* Source/WebCore/platform/graphics/PlatformPlaybackSessionInterface.h:
* Source/WebCore/platform/graphics/PlatformVideoPresentationInterface.h:
Defined PlatformPlaybackSessionInterface to PlaybackSessionInterfaceIOS and
PlatformVideoPresentationInterface to VideoPresentationInterfaceIOS on platforms with
ENABLE(LINEAR_MEDIA_PLAYER). Using abstract base classes allow us to vary the implementation based
on the setting of the LinearMediaPlayer runtime preference.

* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
Added a virtual linearMediaPlayer() and moved m_playerIdentifier from VideoPresentationInterfaceIOS.

* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.mm:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
Added a virtual isExternalPlaybackActive() and moved m_playerIdentifier to PlaybackSessionInterfaceIOS.

* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
Implemented in terms of WKSLinearMediaPlayer.

* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
Created either VideoPresentationInterfaceLMK or VideoPresentationInterfaceAVKit based on the value
of the LinearMediaPlayer runtime preference.

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
Soft-linked WKSLinearMediaPlayer and WKSLinearMediaTimeRange.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
Worked around known issues in Playable.

* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
Used scalars instead of NSValues where possible. Declared -makeViewController.

Canonical link: <a href="https://commits.webkit.org/274743@main">https://commits.webkit.org/274743@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0529ca9d9d6749838a45c84d68e84d90e2fb0fe5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/39889 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/18900 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/42264 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/42433 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35791 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/42196 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/21805 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/16230 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/42433 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/40463 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/21805 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/42264 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/42433 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/21805 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/42264 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/43712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/33342 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/21805 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/42264 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/43712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/39515 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/16230 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/43712 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/16323 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/42264 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/46523 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8949 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16372 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/46523 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
```diff
@@ -1028,6 +1038,7 @@
 			isa = PBXGroup;
 			children = (
 				1C2D0DF8296A44CC00BE33A3 /* libavif.a */,
+				A1D4D22B2B7E04AA0008C40E /* avif_decode_fuzzer */,
```
Unintended change?

aestes replied:
Dunno how that got there. Will revert.

jernoble commented on Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
```diff
@@ -1581,6 +1601,13 @@
+			path = avif_decode_fuzzer;
+			remoteRef = A1D4D22A2B7E04AA0008C40E /* PBXContainerItemProxy */;
+			sourceTree = BUILT_PRODUCTS_DIR;
+		};
```
Ditto.

jernoble commented on Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -26,95 +26,259 @@
+    if (!model)
+        return;
+
+    model->play();
```
This, and a few other places below, could be more succinct:

```
if (auto model = _model.get())
    model->play();
```

aestes committed:
[Cocoa] Implement LinearMediaKit versions of PlaybackSessionInterface and VideoPresentationInterface
https://bugs.webkit.org/show_bug.cgi?id=269442
rdar://problem/122995773

Reviewed by Jer Noble.

Provided implementations of PlaybackSessionInterfaceLMK and VideoPresentationInterfaceLMK backed by
LMPlayableViewController and WKSLinearMediaPlayer.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
Implemented toggleFullscreen by setting the media element's presentation mode.

* Source/WebCore/platform/graphics/PlatformPlaybackSessionInterface.h:
* Source/WebCore/platform/graphics/PlatformVideoPresentationInterface.h:
Defined PlatformPlaybackSessionInterface to PlaybackSessionInterfaceIOS and
PlatformVideoPresentationInterface to VideoPresentationInterfaceIOS on platforms with
ENABLE(LINEAR_MEDIA_PLAYER). Using abstract base classes allow us to vary the implementation based
on the setting of the LinearMediaPlayer runtime preference.

* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
Added a virtual linearMediaPlayer() and moved m_playerIdentifier from VideoPresentationInterfaceIOS.

* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceAVKit.mm:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
Added a virtual isExternalPlaybackActive() and moved m_playerIdentifier to PlaybackSessionInterfaceIOS.

* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.h:
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
Implemented in terms of WKSLinearMediaPlayer.

* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.mm:
Created either VideoPresentationInterfaceLMK or VideoPresentationInterfaceAVKit based on the value
of the LinearMediaPlayer runtime preference.

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
Soft-linked WKSLinearMediaPlayer and WKSLinearMediaTimeRange.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
Worked around known issues in Playable.

* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
Used scalars instead of NSValues where possible. Declared -makeViewController.

Canonical link: https://commits.webkit.org/274743@main
